### ISSUE TITLE
Unmarshal to struct

### ISSUE THREAD

**Issue by burgesQ** [alexejk/go-xmlrpc#84]
Hi, 

I'm trying to unmarshal some xmlrpc response to some go struct without any success 😅 

I'm always end up with either:
- `failed decoding array item at index 2: invalid field type: expected 'struct' or 'map', got 'interface'` 
- `invalid field type: expected 'slice', got 'map'`
- or even fatal failure 

Currently, I got to this: 

```golang
var _payload = `
<methodResponse>
  <params>
    <param>
      <value>
        <array>
          <data>
            <value>
              <i4>200</i4>
            </value>
            <value>OK</value>
            <value>
              <struct>
                <member>
                  <name>status</name>
                  <value>OK</value>
                </member>
                <member>
                  <name>contact</name>
                  <value>&lt;sip:raf@192.168.164.128:5060&gt;;expires=60</value>
                </member>
              </struct>
            </value>
          </data>
        </array>
      </value>
    </param>
  </params>
</methodResponse>`

func TestNewXMLRPC(t *testing.T) {
	resp, err := xmlrpc.NewResponse([]byte(_payload))

	t.Logf("%+v\n", resp.Params)

	require.NoError(t, err)
	dcd := &_xmlrpc.StdDecoder{}
	out := []struct{}{} // any{}
	// struct {
	// 		Test []any // map[string]any //  struct {
	// 		// 	Ret int
	// 		// }
	// 		// // Str []struct {
	// 		// // 	Int int
	// 		// // }
	// 	}{}

	require.NoError(t, dcd.Decode(resp, &out))
	require.Equal(t, &struct{}{}, &out)

	// require.
}
```

A lil help please ? thanks ❤️

**Comment by alexejk** [comment]
Hi, and thanks for raising the question!

What you need is:
```go
out := struct {
		Array []any
	}{}
```

This is because the outer struct wrapper is required, and then the actual response is an array of variable types.
However, you would later hit the problem with struct decoding when embedded into an array. I am opening a PR to solve this shortly. 

Struct will only be decoded into a map then, as it's not possible to provide a strong type for it.

**Comment by alexejk** [comment]
The PR is open and you could try pulling that specific version/commit.
You could try it with: `go get -u alexejk.io/go-xmlrpc@array-mixed-with-structs`

Regarding your test - it's not using a function that this library provides, but modifying it to use e.g StdDecoder directly works just fine from my quick tests.

**Comment by burgesQ** [comment]
> Regarding your test - it's not using a function that this library provides, but modifying it to use e.g StdDecoder directly works just fine from my quick tests.

Correct ! Sorry I lost myself between ur implem, https://github.com/kolo/xmlrpc and https://github.com/divan/gorilla-xmlrpc ! 

Let me re-try w/ ur latest changes from PR

**Comment by alexejk** [comment]
If it works well, i'll merge the PR and release a new version. Let me know.

**Comment by burgesQ** [comment]
### Prior to testing the PR 

The following code fail (removed the embedded struct from the xmlrpc payload for easiness)

```golang
func TestNewXMLRPC(t *testing.T) {
	resp, err := _xmlrpc.NewResponse([]byte(`<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value>
        <array>
          <data>
            <value><i4>200</i4></value>
            <value>OK</value>
          </data>
        </array>
      </value>
    </param>
  </params>
</methodResponse>`))

	require.NoError(t, err)
	dcd := &_xmlrpc.StdDecoder{}
	out := []struct {
		Array []any
	}{}

	require.NoError(t, dcd.Decode(resp, &out))
	require.Equal(t, &struct{}{}, &out)
}
```

<details>
<summary>Traces</summary>

```console
panic: reflect: call of reflect.Value.NumField on slice Value [recovered]
	panic: reflect: call of reflect.Value.NumField on slice Value

goroutine 20 [running]:
testing.tRunner.func1.2({0x632540, 0xc000124ab0})
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1634 +0x377
panic({0x632540?, 0xc000124ab0?})
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/runtime/panic.go:770 +0x132
reflect.flag.mustBe(...)
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/reflect/value.go:233
reflect.Value.NumField({0x624300?, 0xc000124a98?, 0xc000124a98?})
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/reflect/value.go:2127 +0x85
alexejk.io/go-xmlrpc.fieldsMustEqual({0x624340?, 0xc000124a98?}, 0x1)
	/home/master/repo/gopi/src/vendor/alexejk.io/go-xmlrpc/decode.go:305 +0x107
alexejk.io/go-xmlrpc.(*StdDecoder).Decode(0xc0001bdf4f, 0xc000158940, {0x624340, 0xc000124a98})
	/home/master/repo/gopi/src/vendor/alexejk.io/go-xmlrpc/decode.go:46 +0x45
frafos.com/gopi/db/xmlrpc.TestNewXMLRPC(0xc0001576c0)
	/home/master/repo/gopi/src/db/xmlrpc/client_test.go:64 +0xcd
testing.tRunner(0xc0001576c0, 0x6a4d08)
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1742 +0x390

make[1]: *** [../../../make_src/go_test.mk:29: test] Error 1
make[1]: Leaving directory '/home/master/repo/sbc/src/common/api/gopi'
make: [../../../make_src/go_test.mk:47: src/cover/cover.res] Error 2 (ignored)
total:	(statements)	0.0%
```

</details>


### With the PR changes 


Another crash :(   

```console
💀 10% 16:42:29 ❯ head -n 6 src/go.mod                                                                                                                                                           🌿 zsh 🐏 (17%|0%)
module frafos.com/gopi

go 1.22.0

toolchain go1.22.1

require (
	alexejk.io/go-xmlrpc v0.5.2-0.20240325144947-1f615e0c3b67
```

<details>
<summary>Trace</summary>

```console
💀 10% 16:42:29 ❯ make test-cover-re TEST_FILES=./db/xmlrpc                                                                                                                                       make[1]: Entering directory '/home/master/repo/sbc/src/common/api/gopi'
cd ./src && \
GOWORK=off go test -json -shuffle=on -coverpkg ./db/xmlrpc -covermode atomic -coverprofile cover/cover.res ./db/xmlrpc \
2>&1 | tee cover/out.json | gotestfmt
📦 frafos.com/gopi/db/xmlrpc
-test.shuffle 1711381349687947688
  ❌ TestNewXMLRPC (0s)
panic: reflect: call of reflect.Value.NumField on slice Value [recovered]
	panic: reflect: call of reflect.Value.NumField on slice Value

goroutine 19 [running]:
testing.tRunner.func1.2({0x628fa0, 0xc000124708})
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1634 +0x377
panic({0x628fa0?, 0xc000124708?})
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/runtime/panic.go:770 +0x132
reflect.flag.mustBe(...)
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/reflect/value.go:233
reflect.Value.NumField({0x61af20?, 0xc0001246f0?, 0xc0001246f0?})
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/reflect/value.go:2127 +0x85
alexejk.io/go-xmlrpc.fieldsMustEqual({0x61af60?, 0xc0001246f0?}, 0x1)
	/home/master/repo/gopi/src/vendor/alexejk.io/go-xmlrpc/decode.go:316 +0x107
alexejk.io/go-xmlrpc.(*StdDecoder).Decode(0xc0001bff4f, 0xc000158400, {0x61af60, 0xc0001246f0})
	/home/master/repo/gopi/src/vendor/alexejk.io/go-xmlrpc/decode.go:46 +0x45
frafos.com/gopi/db/xmlrpc.TestNewXMLRPC(0xc000157380)
	/home/master/repo/gopi/src/db/xmlrpc/client_test.go:33 +0xcd
testing.tRunner(0xc000157380, 0x69ab10)
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/home/master/repo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.1.linux-amd64/src/testing/testing.go:1742 +0x390

make[1]: *** [../../../make_src/go_test.mk:29: test] Error 1
make[1]: Leaving directory '/home/master/repo/sbc/src/common/api/gopi'
make: [../../../make_src/go_test.mk:47: src/cover/cover.res] Error 2 (ignored)
total:	(statements)	0.0%
```



</details>

**Comment by alexejk** [comment]
The exact response from the original issue is part of PR's test, which passes for me as expected..
I've also added the following test now to validate and it works as expected too

```go
func TestNewXMLRPC2(t *testing.T) {
	resp := struct {
		Array []any
	}{}

	dec := &StdDecoder{}
	require.NoError(t, dec.DecodeRaw([]byte(
		`<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value>
        <array>
          <data>
            <value><i4>200</i4></value>
            <value>OK</value>
          </data>
        </array>
      </value>
    </param>
  </params>
</methodResponse>`), &resp))

	require.Equal(t, &struct{ Array []any }{Array: []any{200, "OK"}}, &resp)
}
```

**Comment by burgesQ** [comment]
Perfect good to know ! Thanks for the quick help. 

For sure the issue is between my keyboard and my chair !

**Comment by alexejk** [comment]
Thanks for helping discover the issue with struct nested in an array. I've merged this and tagged 0.5.2 so you can switch to that version now.
